src/common/src/util/row_id.rs

```diff
@@ -52,35 +55,72 @@ pub struct RowIdGenerator {
 
 pub type RowId = i64;
 
+/// The number of bits occupied by the vnode part of a row id.
+///
+/// In previous versions, this was fixed to 10 bits even if the vnode count was fixed to 256.
+/// For backward compatibility, we still use 10 bits for vnode count less than or equal to 1024.
+/// For larger vnode counts, we use the smallest power of 2 that fits the vnode count.
+fn bit_for_vnode(vnode_count: usize) -> u32 {
+    debug_assert!(
+        vnode_count <= VirtualNode::MAX_COUNT,
+        "invalid vnode count {vnode_count}"
+    );
+
+    if vnode_count <= 1 << COMPAT_VNODE_BITS {
+        COMPAT_VNODE_BITS
+    } else {
+        vnode_count.next_power_of_two().ilog2()
+    }
+}
+
+/// Compute vnode from the given row id.
+///
+/// # `vnode_count`
+///
+/// The given `vnode_count` determines the valid range of the returned vnode. It does not have to
+/// be the same as the vnode count used when the row id was generated with [`RowIdGenerator`].
+///
+/// However, only if they are the same, the vnode retrieved here is guaranteed to be the same as
+/// when it was generated. Otherwise, the vnode can be different and skewed, but the row ids
+/// generated under the same vnode will still yield the same result.
+///
+/// This is okay because we rely on the reversibility only if the serial type (row id) is generated
+/// and persisted in the same fragment, where the vnode count is the same. In other cases, the
+/// serial type is more like a normal integer type, and the algorithm to hash or compute vnode from
+/// it does not matter.
```

Comment:
IIUC, the root cause of the problem is because of the improper way to hack the `hash()` function of `Serial`.

Currently, we did a hack here: if there is only one column and the column type is `Serial`, we will use `extract_vnode_id_from_row_id` instead of the standard hash function.

https://github.com/risingwavelabs/risingwave/blob/8a32a9b21b45c4d91f3235411d35a0a2b3f5102a/src/common/src/hash/consistent_hash/vnode.rs#L133-L168

I think the best solution is to use a special distribution, e.g. `RowIdDistribution` instead of `HashDistribution`. This essentially remove the hack and make everything clear.

cc. @st1page